### PR TITLE
Update to latest JSO APIs

### PR DESCRIPTION
**Pull request by aldma**
This PR brings NCL.jl to the latest JSO interfaces, in particular
- replacing [SolverTools.jl](https://github.com/JuliaSmoothOptimizers/SolverTools.jl) with [SolverCore.jl](https://github.com/JuliaSmoothOptimizers/SolverCore.jl) for [GenericExecutionStats](https://github.com/JuliaSmoothOptimizers/SolverCore.jl/blob/main/src/stats.jl);
- using views when invoking methods of the underlying ``NLPModel``.

Now there is also a list of available solvers, inspired by the one in [CaNNOLeS.jl](https://github.com/JuliaSmoothOptimizers/CaNNOLeS.jl/blob/main/src/CaNNOLeS.jl).

The type change in 20c75610b9cb268f9e25988cd269fe1d9b470c4a (from ``S`` to ``AbstractVector``) is due to julia confusion. This issue should be investigated.

Suitable tests are still missing, as well as a fix for the possible infinite loop due to 

> NCLSolve(nlp::AbstractNLPModel, args...; kwargs...) = NCLSolve(NCLModel(nlp), args...; kwargs...)

and ``NCLModel(nlp)`` returning ``nlp`` untouched.

### CODE REVIEW

**Comment by dpo** [comment]
Many thanks @aldma !!! We need to add CI to this repo. What do you mean by

> Suitable tests are still missing

File: Project.toml [hunk 1]
```diff
@@ -10,7 +10,7 @@ NLPModels = "a4795742-8479-5a88-8948-cc11e1c8c1a6"
 NLPModelsIpopt = "f4238b75-b362-5c4c-b852-0801c9a21d71"
 NLPModelsKnitro = "bec4dd0d-7755-52d5-9a02-22f0ffc7efcb"
 Printf = "de0858da-6303-5e67-8744-51eddeeeb8d7"
-SolverTools = "b5612192-2639-5dc1-abfe-fbedd65fab29"
+SolverCore = "ff4d7338-4cf1-434d-91df-b86cb86fb843"
```
Comment (tmigot) [inline Project.toml:13]: Could you add a compat entry for SolverCore as well?

File: src/NCLModel.jl [hunk 2]
```diff
@@ -105,30 +107,33 @@ end
 function NLPModels.grad!(ncl::NCLModel{T, S, M}, xr::S, gx::S) where {T, S, M <: AbstractNLPModel{T, S}}
   increment!(ncl, :neval_grad)
   x = view(xr, 1 : ncl.nx)
-  grad!(ncl.nlp, x, gx)
+  grad!(ncl.nlp, x, gx[1 : ncl.nx])
```
Comment (tmigot) [inline src/NCLModel.jl]: Should we use view for this?

**Review by tmigot — APPROVED** [review]
Thanks @aldma ! It looks good to me, just two minor comments.

**Comment by aldma** [comment]
The suggestions by @tmigot are integrated now, I hope correctly.
Regarding the tests, I meant that I haven't updated the tests as left to do in [runtests.jl](https://github.com/JuliaSmoothOptimizers/NCL.jl/blob/main/test/runtests.jl).

**Comment by dpo** [comment]
@aldma Please rebase your branch so the CI workflows can run.

**Comment by dpo** [comment]
Please try to use `rebase` in order to avoid merge commits.

**Comment by dpo** [comment]
Superseded by #20.

**Pull request closed without merging**